src/lib/sql/sql_pipeline.cpp

```diff
@@ -0,0 +1,183 @@
+    : _sql(sql), _use_mvcc(true), _auto_commit(false), _transaction_context(std::move(transaction_context)) {
+  DebugAssert(transaction_context != nullptr, "Cannot pass nullptr as explicit transaction context.");
+}
+
+SQLPipeline::SQLPipeline(const std::string& sql, const bool use_mvcc)
+    : _sql(sql), _use_mvcc(use_mvcc), _auto_commit(_use_mvcc) {
+  if (_use_mvcc) {
+    // We want to use MVCC but didn't pass an explicit context
+    _transaction_context = TransactionManager::get().new_transaction_context();
+  } else {
+    // We don't want to use MVCC
+    _transaction_context = nullptr;
+  }
+}
+
+const hsql::SQLParserResult& SQLPipeline::get_parsed_sql() {
+  if (_parsed_sql) {
+    // Returned cached result
+    return *(_parsed_sql.get());
+  }
+
+  auto parse_result = std::make_unique<hsql::SQLParserResult>();
+
+  auto started = std::chrono::high_resolution_clock::now();
+  try {
+    hsql::SQLParser::parse(_sql, parse_result.get());
+  } catch (const std::exception& exception) {
+    throw std::runtime_error("Error while parsing SQL query:\n  " + std::string(exception.what()));
+  }
+
+  if (!parse_result->isValid()) {
+    throw std::logic_error("SQL query not valid.");
+  }
+
+  auto done = std::chrono::high_resolution_clock::now();
+  _parse_time_sec = std::chrono::duration<double>(done - started).count();
+
+  _parsed_sql = std::move(parse_result);
+  return *(_parsed_sql.get());
+}
+
+const std::vector<std::shared_ptr<AbstractLQPNode>>& SQLPipeline::get_unoptimized_logical_plan() {
+  if (!_unopt_logical_plan.empty()) {
+    return _unopt_logical_plan;
+  }
+
+  const auto& parsed_sql = get_parsed_sql();
+  try {
+    _unopt_logical_plan = SQLTranslator{_use_mvcc}.translate_parse_result(parsed_sql);
+  } catch (const std::exception& exception) {
+    throw std::runtime_error("Error while compiling query plan:\n  " + std::string(exception.what()));
+  }
+
+  return _unopt_logical_plan;
+}
+
+const std::vector<std::shared_ptr<AbstractLQPNode>>& SQLPipeline::get_optimized_logical_plan() {
+  if (!_opt_logical_plan.empty()) {
+    return _opt_logical_plan;
+  }
+
+  auto unopt_lqp = get_unoptimized_logical_plan();
+  _opt_logical_plan.reserve(unopt_lqp.size());
+
+  try {
+    for (const auto& node : unopt_lqp) {
+      _opt_logical_plan.push_back(Optimizer::get().optimize(node));
+    }
+  } catch (const std::exception& exception) {
+    throw std::runtime_error("Error while optimizing query plan:\n  " + std::string(exception.what()));
+  }
+
+  return _opt_logical_plan;
+}
+
+const SQLQueryPlan& SQLPipeline::get_query_plan() {
+  if (_query_plan) {
+    return *(_query_plan.get());
+  }
+
+  const auto& lqp_roots = get_optimized_logical_plan();
+  auto plan = std::make_unique<SQLQueryPlan>();
+
+  auto started = std::chrono::high_resolution_clock::now();
+
+  try {
+    for (const auto& node : lqp_roots) {
+      auto optimized = Optimizer::get().optimize(node);
```

Comment:
You optimize this... again?

Reply:
Ah, this is so inserted View get optimized as well or something like that? In any case, please doc

Reply:
This seems like a copy-paste error. We assume the LQP is optimized already.